From 60719453e0c09ad04a4a047efd815ec04d71cfd7 Mon Sep 17 00:00:00 2001
From: Tanuj Tekriwal <tanuj.tekriwal@intel.com>
Date: Wed, 22 Sep 2021 17:38:58 +0530
Subject: [PATCH] [WA] Aquire wakelock when BluetoothManagerService retries
 enable BT

com.android.bluetooth process crashes on system resume if power is
cut to the bluetooth chip during suspend. If system suspends again
before bluetooth is enabled, the crash repeats, and which will result
in the subsequent retries to fail. After many suspsed resume cycles
the retry counter will exceed the maximum limit and BT will remain in
Off state.

Workaround solution is to acquire wakelock when the service tries to
restart BT

Change-Id: Ib696f064fc9123350fb7a6093d9a15aaa2a843c6
Tracked-On: OAM-97107
Signed-off-by: Aiswarya Cyriac <aiswarya.cyriac@intel.com>
---
 .../android/server/BluetoothManagerService.java | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/services/core/java/com/android/server/BluetoothManagerService.java b/services/core/java/com/android/server/BluetoothManagerService.java
index b7c61a0d25cf..0c395b34b18d 100644
--- a/services/core/java/com/android/server/BluetoothManagerService.java
+++ b/services/core/java/com/android/server/BluetoothManagerService.java
@@ -67,6 +67,8 @@ import android.os.Handler;
 import android.os.IBinder;
 import android.os.Looper;
 import android.os.Message;
+import android.os.PowerManager;
+import android.os.PowerManager.WakeLock;
 import android.os.PowerExemptionManager;
 import android.os.Process;
 import android.os.RemoteCallbackList;
@@ -189,6 +191,8 @@ class BluetoothManagerService extends IBluetoothManager.Stub {
     private boolean mUnbinding;
     private int mWaitForEnableRetry;
     private int mWaitForDisableRetry;
+    private PowerManager mPowerManager;
+    private WakeLock mPartialWakeLock;
 
     private BluetoothModeChangeHelper mBluetoothModeChangeHelper;
 
@@ -503,6 +507,7 @@ class BluetoothManagerService extends IBluetoothManager.Stub {
         mIsHearingAidProfileSupported = context.getResources()
                 .getBoolean(com.android.internal.R.bool.config_hearing_aid_profile_supported);
 
+        mPowerManager = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
         // TODO: We need a more generic way to initialize the persist keys of FeatureFlagUtils
         String value = SystemProperties.get(FeatureFlagUtils.PERSIST_PREFIX + FeatureFlagUtils.HEARING_AID_SETTINGS);
         if (!TextUtils.isEmpty(value)) {
@@ -555,6 +560,8 @@ class BluetoothManagerService extends IBluetoothManager.Stub {
             Slog.w(TAG, "Unable to resolve SystemUI's UID.");
         }
         mSystemUiUid = systemUiUid;
+        mPartialWakeLock = mPowerManager.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, TAG);
+        mPartialWakeLock.setReferenceCounted(false);
     }
 
     /**
@@ -2166,6 +2173,9 @@ class BluetoothManagerService extends IBluetoothManager.Stub {
                             Slog.w(TAG, "bluetooth is recovered from error");
                             mErrorRecoveryRetryCounter = 0;
                         }
+                        if (mPartialWakeLock.isHeld()) {
+                            mPartialWakeLock.release();
+                        }
                     }
                     break;
                 }
@@ -2223,6 +2233,9 @@ class BluetoothManagerService extends IBluetoothManager.Stub {
                 }
                 case MESSAGE_RESTART_BLUETOOTH_SERVICE: {
                     mErrorRecoveryRetryCounter++;
+                    if (!mPartialWakeLock.isHeld()) {
+                        mPartialWakeLock.acquire();
+                    }
                     Slog.d(TAG, "MESSAGE_RESTART_BLUETOOTH_SERVICE: retry count="
                             + mErrorRecoveryRetryCounter);
                     if (mErrorRecoveryRetryCounter < MAX_ERROR_RESTART_RETRIES) {
@@ -2235,6 +2248,10 @@ class BluetoothManagerService extends IBluetoothManager.Stub {
                         handleEnable(mQuietEnable);
                     } else {
                         Slog.e(TAG, "Reach maximum retry to restart Bluetooth!");
+                        if (mPartialWakeLock.isHeld()) {
+                            Slog.d(TAG,"Releasing the partial wakelock as maximum retry exceeded");
+                            mPartialWakeLock.release();
+                        }
                     }
                     break;
                 }
-- 
2.17.1

